src/Model/Table/SentencesTable.php

```diff
@@ -1259,7 +1259,16 @@ public function deleteSentence($id)
             return false;
         }
 
-        return $this->delete($sentence);
+        $listsContainingSentence = $this->loadInto($sentence, ['SentencesLists'])->sentences_lists;
+        $isDeleted = $this->delete($sentence);
+        if ($isDeleted) {
+            foreach ($listsContainingSentence as $list){
+                $list->numberOfSentences--;
+                $this->SentencesLists->save($list);
+            }
+        }
+
+        return $isDeleted;
```

Comment:
This code works, but it can be improved. First, you can move it to the `afterDelete()` method, which will only be executed if the deletion succeeds, so you don’t have to carry around `$isDeleted`. That would be, let’s say, improvement A.

As you can see, `afterDelete()` is already quite busy doing a lot of things which are not of SentencesTable’s responsibility. Updating the number of sentences in the lists is also not something SentencesTable should have to care about. So we can improve this even more by moving your code into the SentencesLists class and have it executed with some kind of event/triggering system. That would be improvement B, and it’s super good because it reduces [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) between SentencesTable and SentencesLists. For an example of how to do that, look at `SentencesTable::afterSave()` and `src/Event/ContributionListener.php` (and the unit tests).

Consider improvement B as a bonus. I will still merge your PR if you only do improvement A.

Reply:
Now, I remember why I wrote it like this. If done after deletion, the sentence is already gone so there are no associations between this sentence and the corresponding sentencesLists, hence rows to decrement cannot be fetched.   
If we check the `sentencesList.ids` before deletion, and then pass them to decrement them, might as well directly do it. That's what I thought.

Now that I think again, maybe the easiest trigger is not on `sentences` but on `sentences_sentences_lists`? Because we would have the `list_id` available directly. But I don't know if a similar behavior is possible or how to do it. By that I mean that `sentence` deletion cascades to `sentences_sentences_list` deletion, and that one could cascade to `sentences_list` decrement. 

Any advice? Is it correct to have an afterDelete on the intermediate table?

Reply:
I see. Yes, putting your code in the link table model sounds like a good idea, if it works. You’d have to create a method named `afterDelete()` in SentencesSentencesListsTable using the same signature as in SentencesTable. The deleted link should be provided as a second parameter of `afterDelete()`.

Reply:
I had to do a [little research](https://book.cakephp.org/3/en/orm/deleting-data.html#cascading-deletes) about that. It seems that we need to set the `cascadeCallbacks` option of the association to `true` to be able to propagate the call to `afterDelete()` (and other callbacks) of the junction table. The documentation warns us 

> Setting cascadeCallbacks to true, results in considerably slower deletes when compared to bulk deletes. The cascadeCallbacks option should only be enabled when your application has important work handled by event listeners.

but I think it's okay because we don't delete sentences so much and there aren't many associations to delete.